### PR TITLE
refactor(types): instance.ts

### PR DESCRIPTION
**Pull request by joecervino**
Fixes #<issue_number_goes_here> (it's a good idea to open an issue first for discussion)

- [ ] Tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

### CODE REVIEW

File: src/instance.ts [hunk 1]
```diff
@@ -629,7 +643,7 @@ class Instance extends common.ServiceObject {
    *   const apiResponse = data[1];
    * });
    */
-  setMetadata(metadata, callback?) {
+  setMetadata(metadata: Any, callback?: Function): void {
```
Comment (nareshqlogic) [inline src/instance.ts]: If callback is option then it should return  Promise<LongRunningOperationResponse> and callback should be type LongRunningOperationCallback

File: src/instance.ts [hunk 2]
```diff
@@ -574,7 +588,7 @@ class Instance extends common.ServiceObject {
    *   const apiResponse = data[1];
    * });
    */
-  getMetadata(callback) {
+  getMetadata(callback: Function): void {
```
Comment (nareshqlogic) [inline src/instance.ts]: Here callback should be is optional as mentioned in function comment and it should return void|Promise<GetInstanceMetadataResponse>

File: src/instance.ts [hunk 3]
```diff
@@ -415,30 +430,29 @@ class Instance extends common.ServiceObject {
    *   const apiResponse = data[0];
    * });
    */
-  get(options, callback?) {
-      if (is.fn(options)) {
-        callback = options;
-        options = {};
-      }
-      this.getMetadata((err, metadata) => {
+  get(options: Any, callback?: Function): void {
```
Comment (nareshqlogic) [inline src/instance.ts]: Can you please change type from any to proper type for the options? and callback must be  type GetInstanceCallback and it should return void| Promise<GetInstanceResponse>

File: src/instance.ts [hunk 4]
```diff
@@ -359,17 +374,17 @@ class Instance extends common.ServiceObject {
    *   const exists = data[0];
    * });
    */
-  exists(callback) {
+  exists(callback: Function): void {
```
Comment (nareshqlogic) [inline src/instance.ts]: Here callback should be optional as mentioned in function comment and it should be type of InstanceExistsCallback and function should return void |Promise<InstanceExistsResponse>

File: src/instance.ts [hunk 5]
```diff
@@ -415,30 +430,29 @@ class Instance extends common.ServiceObject {
...
    */
-  get(options, callback?) {
-      if (is.fn(options)) {
-        callback = options;
-        options = {};
-      }
-      this.getMetadata((err, metadata) => {
+  get(options: Any, callback?: Function): void {
+      callback = typeof options === 'function' ? options : callback!;
+      options = typeof options === 'object' ? options : {autoCreate: false};
```
Comment (nareshqlogic) [inline src/instance.ts]: Are you sure default value of options should be {autoCreate: false}?

File: src/instance.ts [hunk 6]
```diff
@@ -43,7 +56,7 @@ import {Database} from './database';
  * const instance = spanner.instance('my-instance');
  */
 class Instance extends common.ServiceObject {
-  constructor(spanner, name) {
+  constructor(spanner: Any, name: string) {
```
Comment (nareshqlogic) [inline src/instance.ts]: Here spanner should be Spanner type.

File: src/instance.ts [hunk 7]
```diff
@@ -101,7 +114,7 @@ class Instance extends common.ServiceObject {
        */
       id: name,
       methods,
-      createMethod(_, options, callback) {
+      createMethod(_: Any, options: Any, callback: Function) {
```
Comment (nareshqlogic) [inline src/instance.ts]: Please change type for _ to {}

File: src/instance.ts [hunk 8]
```diff
@@ -22,9 +22,22 @@ import {ServiceObjectConfig} from '@google-cloud/common';
...
+import * as r from 'request';
+
+export interface CreateDatabaseOptions {
+  [key: string]: Any;
+  poolOptions?: {};
+  schema?: string;
+  poolCtor?: {};
+}
+export type CreateDatabaseResponse = [Any, ("insert"|"update"|"insertOrUpdate"|"replace"|"delete"), r.Response];
+export interface ExtendedError extends Error {
```
Comment (praveenqlogic) [inline src/instance.ts]: better! if we keep types in index.ts file!

File: src/instance.ts [hunk 9]
```diff
@@ -194,44 +207,47 @@ class Instance extends common.ServiceObject {
...
    * Full example:
    */
-  createDatabase(name, options?, callback?) {
+  createDatabase(name: string, options?: CreateDatabaseOptions): Promise<CreateDatabaseResponse>;
+  createDatabase(name: string, callback: Function): void;
+  createDatabase(name: string, options: CreateDatabaseOptions, callback: Function): void;
+  createDatabase(
+    name: string,
+    optionsOrCallback?: (CreateDatabaseOptions|Function),
+    cb?: Function): (void|Promise<CreateDatabaseResponse>) {
```
Comment (praveenqlogic) [inline src/instance.ts]: below type can help for callback!
export interface CreateDatabaseCallback{
  (err:ExtendedError|null,database:Any,operation:Operations|null,resp: r.Response):void;  
}

File: src/instance.ts [hunk 10]
```diff
@@ -301,14 +317,13 @@ class Instance extends common.ServiceObject {
    *   const apiResponse = data[0];
    * });
    */
-  delete(callback) {
+  delete(callback: Function): void {
```
Comment (praveenqlogic) [inline src/instance.ts]: delete():Promise&lt;google.protobuf.Empty&gt;;
delete(callback:DeleteInstanceCallback): void;
delete(callback?:DeleteInstanceCallback): void | Promise&lt;google.protobuf.Empty&gt;

export interface DeleteInstanceCallback{
  (err:ExtendedError|null,resp?:google.protobuf.Empty|null):void;
}

File: src/instance.ts (hunk 3 above)
Comment (praveenqlogic) [reply src/instance.ts]: you can use below code!

get(options: GetInstanceOptions):Promise&lt;Instance&gt;;
  get(options: GetInstanceOptions, callback:GetInstanceCallback): void; 
  get(optionsOrCallback?: GetInstanceOptions | GetInstanceCallback, callback?:GetInstanceCallback): void | Promise&lt;Instance&gt; {



types////
import {CallOptions} from 'google-gax';
export interface GetInstanceOptions extends CallOptions{
  autoCreate:boolean;
}

export interface GetInstanceCallback{
  (err:Error|null,instance?:Instance|null,resp?:object|null):void;
}

File: src/instance.ts (hunk 5 above)
Comment (praveenqlogic) [reply src/instance.ts]: const options = (typeof optionsOrCallback === 'object' ? optionsOrCallback : {}) as GetInstanceOptions;
    callback = typeof optionsOrCallback === 'function' ? optionsOrCallback : callback;

File: src/instance.ts (hunk 2 above)
Comment (praveenqlogic) [reply src/instance.ts]: you can follow below lines

getMetadata():Promise&lt;Instance&gt;;
  getMetadata(callback: GetInstanceMetadataCallback):void;
  getMetadata(callback?: GetInstanceMetadataCallback): void | Promise&lt;Instance&gt; {

File: src/instance.ts [hunk 11]
```diff
@@ -508,12 +522,12 @@ class Instance extends common.ServiceObject {
    *   const databases = data[0];
    * });
    */
-  getDatabases(query, callback?) {
+  getDatabases(query: (string[]|{}), callback?: Function): Any {
```
Comment (praveenqlogic) [inline src/instance.ts]: here also we can create callback below might work :)
 (err?: Error|null, databases?: Database[]|null,
   apiResponse?: object): void;

File: src/instance.ts (hunk 1 above)
Comment (JCIV-SE) [reply src/instance.ts]: Okay! Thanks!

File: src/instance.ts (hunk 2 above)
Comment (JCIV-SE) [reply src/instance.ts]: Thanks guys!

File: src/instance.ts (hunk 4 above)
Comment (JCIV-SE) [reply src/instance.ts]: I really shouldn't have decided to disregard the documentation, seems like that's exactly what I was supposed to follow. Thanks guys!

File: src/instance.ts (hunk 6 above)
Comment (JCIV-SE) [reply src/instance.ts]: Using the Spanner type produces an error that states there are ~27 missing parameters 😄

File: src/instance.ts (hunk 9 above)
Comment (JCIV-SE) [reply src/instance.ts]: Thanks!

File: src/instance.ts (hunk 10 above)
Comment (JCIV-SE) [reply src/instance.ts]: Thanks!!

File: src/instance.ts (hunk 11 above)
Comment (JCIV-SE) [reply src/instance.ts]: Thanks Praveen!

File: src/instance.ts (hunk 5 above)
Comment (JCIV-SE) [reply src/instance.ts]: Done! Thanks!

File: src/instance.ts (hunk 3 above)
Comment (JCIV-SE) [reply src/instance.ts]: Sweet! Done!

File: src/instance.ts [hunk 12]
```diff
@@ -22,9 +22,22 @@ import {ServiceObjectConfig} from '@google-cloud/common';
 import {paginator} from '@google-cloud/paginator';
```
[the remark this answers is not in the document]
Comment (praveenqlogic) [reply src/instance.ts]: i am not sure on it! but as per me, either than using any type is SRC its better to use in test files! and that fix lot of bugs in your file, you can use spanner type also for instance class constructor!

**Pull request closed without merging**